backend/models/sequenceActivities.js

```diff
@@ -0,0 +1,298 @@
+	]);
+
+	return {...activity, _id, sequenceId: sequenceId};
+};
+
+SequenceActivities.edit = async (account, model, sequenceId, activityId, activity) => {
+	const sequence = await db.findOne(account, model + ".sequences", { _id: utils.stringToUUID(sequenceId)});
+
+	if (!sequence) {
+		throw responseCodes.SEQUENCE_NOT_FOUND;
+	}
+
+	if (!activityEditSchema.isValidSync(activity, { strict: true })) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	if (activity.parent && !await db.findOne(account, activityCol(model), { _id: utils.stringToUUID(activity.parent)})) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	const query =  {_id: utils.stringToUUID(activityId), sequenceId: utils.stringToUUID(sequenceId)};
+	const {result} = await db.update(account, activityCol(model), query, {$set: activity});
+
+	if (!result.n) {
+		throw responseCodes.ACTIVITY_NOT_FOUND;
+	}
+
+	await FileRef.removeFile(account, model, "activities", sequenceId);
+};
+
+SequenceActivities.remove = async (account, model, sequenceId, activityId) => {
+	const sequence = await db.findOne(account, model + ".sequences", { _id: utils.stringToUUID(sequenceId)});
+
+	if (!sequence) {
+		throw responseCodes.SEQUENCE_NOT_FOUND;
+	}
+	const activities = await db.find(account,activityCol(model), {sequenceId: utils.stringToUUID(sequenceId), parent:{$exists: true}} , {parent:1, _id:1});
+
+	const idsToDelete = getDescendantsIds(activities, activityId);
+	idsToDelete.push(utils.stringToUUID(activityId));
+
+	// console.log(idsToDelete.map(utils.uuidToString));
+
+	const query = {_id:{ $in: idsToDelete}, sequenceId: utils.stringToUUID(sequenceId)};
+	const {result} = await db.remove(account,  activityCol(model), query);
+
+	if (!result.n) {
+		throw responseCodes.ACTIVITY_NOT_FOUND;
+	}
+
+	await FileRef.removeFile(account, model, "activities", sequenceId);
+
+	return { _id: activityId};
+};
+
+SequenceActivities.get = async (account, model, sequenceId) => {
+	const sequence = await db.findOne(account, model + ".sequences", { _id: utils.stringToUUID(sequenceId)});
+
+	if (!sequence) {
+		throw responseCodes.SEQUENCE_NOT_FOUND;
+	}
+
+	let activities = {};
+	try {
+		activities = await FileRef.getSequenceActivitiesFile(account, model, utils.uuidToString(sequenceId));
+	} catch(e) {
+		activities = await createActivitiesTree(account, model, sequenceId);
+		await FileRef.storeFile(account, activityCol(model) + ".ref", account, utils.uuidToString(nodeuuid()), JSON.stringify(activities),  {"_id": sequenceId});
+	}
+
+	return activities;
+};
+
+SequenceActivities.createActivities = async (account, model, sequenceId, activitiesTree, overwrite) => {
+	const sequence = await db.findOne(account, model + ".sequences", { _id: utils.stringToUUID(sequenceId)});
+
+	if (!sequence) {
+		throw responseCodes.SEQUENCE_NOT_FOUND;
+	}
+
+	if (!activityTreeSchema.isValidSync(activitiesTree, { strict: true })) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	fillIds(activitiesTree);
```

Comment:
The activityTreeSchema check at L276 and this line means we're traversing the tree twice.

I think it's better if we just rewrite it to traverse the tree once, and do both the schema check and append IDs